libproxy/libproxyd.cpp

```diff
@@ -0,0 +1,177 @@
+      {
+          g_variant_builder_add(result, "&s", proxies[j]);
+	  free(proxies[j]);
+      }
+      g_free (proxies);
+
+      g_dbus_method_invocation_return_value (invocation,
+                                             g_variant_new("(as)", result));
+    }
+}
+
+static GVariant *
+handle_get_property (GDBusConnection  *connection,
+                     const gchar      *sender,
+                     const gchar      *object_path,
+                     const gchar      *interface_name,
+                     const gchar      *property_name,
+                     GError          **error,
+                     gpointer          user_data)
+{
+  GVariant *ret;
+
+  ret = NULL;
+  if (g_strcmp0 (property_name, "APIVersion") == 0)
+    {
+      ret = g_variant_new_string ("1.0");
+    }
+
+  return ret;
+}
+
+/* for now */
+static const GDBusInterfaceVTable interface_vtable =
+{
+  handle_method_call,
+  handle_get_property
+};
+
+/* ---------------------------------------------------------------------------------------------------- */
+
+static void
+on_bus_acquired (GDBusConnection *connection,
+                 const gchar     *name,
+                 gpointer         user_data)
+{
+  guint registration_id;
+
+  registration_id = g_dbus_connection_register_object (connection,
+                                                       "/org/libproxy/proxy",
+                                                       introspection_data->interfaces[0],
+                                                       &interface_vtable,
+                                                       NULL,  /* user_data */
+                                                       NULL,  /* user_data_free_func */
+                                                       NULL); /* GError** */
+  g_assert (registration_id > 0);
+
+}
+
+static void
+on_name_acquired (GDBusConnection *connection,
+                  const gchar     *name,
+                  gpointer         user_data)
+{
+}
+
+static void
+on_name_lost (GDBusConnection *connection,
+              const gchar     *name,
+              gpointer         user_data)
+{
+  exit (1);
+}
+
+int
+main (int argc, char *argv[])
+{
+  guint owner_id;
+  GMainLoop *loop;
+
+  /* We are lazy here - we don't want to manually provide
+   * the introspection data structures - so we just build
+   * them from XML.
+   */
+  introspection_data = g_dbus_node_info_new_for_xml (introspection_xml, NULL);
+  g_assert (introspection_data != NULL);
+
+  owner_id = g_bus_own_name (G_BUS_TYPE_SESSION,
+                             "org.libproxy.proxy",
+                             G_BUS_NAME_OWNER_FLAGS_NONE,
+                             on_bus_acquired,
+                             on_name_acquired,
+                             on_name_lost,
+                             NULL,
+                             NULL);
+
+  pf = px_proxy_factory_new();
+
+  loop = g_main_loop_new (NULL, FALSE);
+  g_main_loop_run (loop);
```

Comment:
Actually, maybe the daemon should not run forever. Since it is a D-Bus service, it should probably terminate after, say, one minute of inactivity. Easy way to do that would be to use GApplication, pass it an inactivity timeout, and call g_application_hold()/g_application_release() in the query method call handler. After all, a typical desktop user will never use libproxy at all, so no need for it to be running unnecessarily.

Reply:
actually, the daemon will run on most systems probably forever; libproxy is even queried when no proxy is configured - that's how apps know that know proxy needs to be used.

Reply:
> libproxy is even queried when no proxy is configured - that's how apps know that know proxy needs to be used

True, but still not a great reason for it to run forever. E.g. if the user is using GNOME, it's likely nothing will ever query libproxy, since GNOME apps will just check GNOME proxy settings directly, so only apps that link directly to libproxy will wind up using it. There's also https://gitlab.gnome.org/GNOME/glib-networking/-/issues/19, which is a second potential way libproxy might be cut out of deciding which proxy gets used in the future. Or maybe no apps are using the network!

Reply:
imho a fault of GNOME to cut out libproxy: gnome-apps should use the proxy configured in KDE when running in a KDE session; and gnome code knowing how to read KDE settings sounds more than strange.

This is what libproxy is solving after all

Reply:
Well, that's what we do. We only read GNOME settings when "GNOME" is present in XDG_CURRENT_DESKTOP. Otherwise, we fall back to libproxy and hope it doesn't crash. :P

Reply:
:) and that is what we try to change: no more 'hoping'

I appreciate your knowledge/help to move in this direction. Let's try to get this dbusify branch ready soon™